### PR TITLE
Form Storing Emails

### LINKED ISSUE #1463: Form Storing Emails
### Description:

In Forms, we want to allow users to store submitted emails on their WordPress so they can be accessed if they want. To do this, we can use the actions available in Otter Forms instead of hardcoding the code, I think `otter_form_after_submit` should be the one but one can confirm before implementation.

We can simply use a CPT to store the data without its UI, and display it inside Otter Dashboard in a new subpage (an actual subpage). While doing this, we can [move Otter page to a top-level menu item](https://github.com/Codeinwp/otter-blocks/issues/575) with Settings & Form Responses as their menus.

The option should be true by default but the user can disable if needed. Also, we can use the same page structure as the existing Dashboard with just simple Accordion-style elements to display the responses (as well as the option to delete them). An example UI being this: https://tripetto.com/images/help/wordpress-results/result.png

Feel free to ask for more details if needed.

### PR DESCRIPTION
**Pull request by arinaturcu**
Closes #1463, #575.

### Summary

- Creates a new top-level menu for Otter with Settings and Form Submissions as pages.
- All form submissions from the site are stored on the Form Submissions page.
- Each form submission is of a custom post type named `otter_form_record` and the data from the input fields, the submission date, and the form ID are saved as metadata with the key `otter_form_record_meta`.
- Each submission can be edited, deleted, or marked as read/unread. Creating a new submission is not allowed.
- Submission can be filtered by month, page, or form.
- A submission is marked as `read` when it is viewed (by clicking `View`) or edited.
- It is a pro feature, so if Otter Pro is not activated, the page will still be visible in the menu but with placeholder content as in the images below.
- Only the administrator has access to these form submissions.

### Screenshots 
Screenshot 2023-04-30 at 00 27 54

Screenshot 2023-04-30 at 00 29 14

Screenshot 2023-04-30 at 00 27 25

----

### Test instructions

- Add form blocks in various places on the site, and even multiple forms on one page.
- Make different submissions and make sure all the features described above work as expected.
- Make sure that user roles other than `administrator` don't have access to see/edit/delete the form submissions. 

---- 

### Checklist before the final review

- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.
- [x] Copy/Paste is working if the attributes are modified.

### CODE REVIEW

File: inc/plugins/class-dashboard.php [hunk 1]
```diff
@@ -8,6 +8,7 @@
 namespace ThemeIsle\GutenbergBlocks\Plugins;
 
 use ThemeIsle\GutenbergBlocks\Pro;
+use function Sodium\add;
```
Comment (HardeepAsrani) [inline inc/plugins/class-dashboard.php:11]: @arinaturcu is this added by VSCode's autocompletion?

**Review by HardeepAsrani — CHANGES_REQUESTED** [review]
For some reason, I see these two Pro areas even while using the Pro version:

Screenshot 2023-05-03 at 1 04 59 AM

For now, we don't have any option to disable the form submission? Just asking, we can add that option later when Robert's Form work is also done to avoid conflicts.

File: plugins/otter-pro/inc/plugins/form/class-form-submissions.php [hunk 2]
```diff
@@ -0,0 +1,983 @@
...
+	 * Set form records bulk actions.
+	 *
+	 * @return array
+	 */
+	public function form_record_bulk_actions() {
+		$status       = isset( $_GET['post_status'] ) ? sanitize_text_field( wp_unslash( $_GET['post_status'] ) ) : 'all'; // phpcs:ignore WordPress.Security.NonceVerification.NoNonceVerification
+		$bulk_actions = array();
+
+		if ( 'trash' !== $status ) {
+			$bulk_actions['trash'] = 'Move to Trash';
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/form/class-form-submissions.php:283]: Here you need to use i18n functions.

File: plugins/otter-pro/inc/plugins/form/class-form-submissions.php [hunk 3]
```diff
@@ -0,0 +1,983 @@
...
+		add_action( 'save_post', array( $this, 'form_record_save_meta_box' ), 10, 2 );
+	}
+
+	/**
+	 * Create custom post type for form records.
+	 *
+	 * @return void
+	 */
+	public function create_form_records_type() {
+		register_post_type(
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/form/class-form-submissions.php:79]: One major issue right now that I see is that any user even with the role of Contributor can access this Dashboard. We should limit it to only people with admin capabilities.

Comment (HardeepAsrani) [reply plugins/otter-pro/inc/plugins/form/class-form-submissions.php:79]: If you're using a role lower than Admin, you can visit the Dashboard by accessing the URl directly.

**Comment by arinaturcu** [comment]
Thanks, @HardeepAsrani!
I fixed those and gave capabilities only to the administrator role. Let me know if you find anything else.

**Review by HardeepAsrani — APPROVED** [review]
Looks pretty great.

**Comment by rodica-andronache** [comment]
@arinaturcu I've used the development versions https://vertis.d.pr/i/8sUE5X ( let me know if that's ok )

For Otter (free) I noticed that the form submissions page is not responsive https://vertis.d.pr/i/x7974m as is the main Otter page

When installing Otter Pro I got a fatal error https://vertis.d.pr/i/xWZffT so I've put the testing on hold

**Comment by arinaturcu** [comment]
Thanks @rodica-andronache! It should work fine now 😄
Also, the free form submissions page is responsive now.

**Comment by rodica-andronache** [comment]
@arinaturcu here are my findings. Thank you!

- [ ] Click on View submission -> than try to **Move to trash** -> Is not working https://vertis.d.pr/v/dZqnE8

- [ ] If the Email **Label** is empty ( and I’m using just the **Placeholder** for example ), the submission is not recorded  https://vertis.d.pr/i/oniQGF

- [ ] Similar for other fields, for e.g if I have a Textarea without a Label ( using just the Placeholder ), that field is not registered in the submission

- [ ] After a submission is created, if you go ahead and change the **Permalinks**, the link for the Post ( From where the form submission is coming from ) is not working anymore ( as it doesn’t get updated based on the Permalinks ) https://vertis.d.pr/v/8Vq1mR

- [ ] The fields from the submissions are not that responsive https://vertis.d.pr/i/AN3NTv ( compared to the PODs plugin for example )

**Comment by arinaturcu** [comment]
@HardeepAsrani I see that the issue found by Rodica comes from here, and seems intentional. Is this right, or can I just save the input with `'(No label)'` as the label? What is the reason behind checking it?
https://github.com/Codeinwp/otter-blocks/blob/d4da09f67f35ebfc4f818a997601552e0a8f713b/src/blocks/frontend/form/index.js#L37-L45

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel do you have any context to add here?

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel do you have any context to add here?

I did not consider fields without a Label as valid fields. 

@arinaturcu You can remove the `&& valueElem?.value` and add a default value.

File: plugins/otter-pro/inc/plugins/class-form-emails-storing.php [hunk 4]
```diff
@@ -0,0 +1,983 @@
...
+			return;
+		}
+
+		$post_id = wp_insert_post(
+			array(
+				'post_type'   => self::FORM_RECORD_TYPE,
+				'post_title'  => $email,
+				'post_status' => 'unread',
+			)
+		);
```
[the remark this answers is not in the document]
Comment (arinaturcu) [reply plugins/otter-pro/inc/plugins/class-form-emails-storing.php:183]: This sounds really good 👍🏻

**Comment by Soare-Robert-Daniel** [comment]
@Codeinwp/qa-team, for Otter this PR will be a priority 🙏 

Things added:
- Choose save location: Database (the Submissions menu), Email Only, Database and Email
- File field integration: You should see links and images when you submit some files.

:warning: When upgrading to Otter Pro, the email will set to be saved on submission instead of keep being sent as email.

⚠️ Functionality must be a priority. Things should work as expected. Details regarding stylistics are non-blocker; you can mention them, but depending on complexity they might not be addressed on this PR.

🔴 This will go through multiple rounds of testing.

## Screenshots

### Choose Location

image

### Files saved on submission

image

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested and the issues mentioned by Rodica are fixed, thank you 🚀 

I have only one question:

If you have multiple files attached in the form (3 in my example), there will be 4 entries in the database https://vertis.d.pr/i/W77Tpq . Is this expected?

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, it is ok if you used Save to the Media Library. What you see are the attachments generated for the images.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Thank you for checking 🙏 I'll move this to **Ready to merge** as everything's fine here 🚀

**Pull request merged**